README.md
ZF Api Problem
==============

[![Build Status](https://secure.travis-ci.org/zfcampus/zf-api-problem.svg?branch=master)](https://secure.travis-ci.org/zfcampus/zf-api-problem)
[![Coverage Status](https://coveralls.io/repos/github/zfcampus/zf-api-problem/badge.svg?branch=master)](https://coveralls.io/github/zfcampus/zf-api-problem?branch=master)

Introduction
------------

This module provides data structures and rendering for the API-Problem format.

- [Problem Details for HTTP APIs](https://tools.ietf.org/html/rfc7807),
  used for reporting API problems.

Requirements
------------
  
Please see the [composer.json](composer.json) file.

Installation
------------

Run the following `composer` command:

```console
$ composer require zfcampus/zf-api-problem
```

Alternately, manually add the following to your `composer.json`, in the `require` section:

```javascript
"require": {
    "zfcampus/zf-api-problem": "^1.2"
}
```

And then run `composer update` to ensure the module is installed.

Finally, add the module name to your project's `config/application.config.php` under the `modules`
key:

```php
return [
    /* ... */
    'modules' => [
        /* ... */
        'ZF\ApiProblem',
    ],
    /* ... */
];
```

> ### zf-component-installer
>
> If you use [zf-component-installer](https://github.com/zendframework/zf-component-installer),
> that plugin will install zf-api-problem as a module for you.

Configuration
=============

### User Configuration

The top-level configuration key for user configuration of this module is `zf-api-problem`.

#### Key: accept_filters

An array of `Accept` header media types that, when matched, will result in the
[ApiProblemListener](#zfapiproblemlistenerapiproblemlistener) handling an
`MvcEvent::EVENT_RENDER_ERROR` event.

#### Key: render_error_controllers

An array of controller service names that, if matched as the `controller` parameter in the MVC
`RouteMatch`, will cause the [ApiProblemListener](#zfapiproblemlistenerapiproblemlistener) to handle 
`MvcEvent::EVENT_RENDER_ERROR` events.

### System Configuration

The following configuration is provided in `config/module.config.php` to enable the module to
function:

```php
'service_manager' => [
    'aliases'   => [
        'ZF\ApiProblem\ApiProblemListener'  => 'ZF\ApiProblem\Listener\ApiProblemListener',
        'ZF\ApiProblem\RenderErrorListener' => 'ZF\ApiProblem\Listener\RenderErrorListener',
        'ZF\ApiProblem\ApiProblemRenderer'  => 'ZF\ApiProblem\View\ApiProblemRenderer',
        'ZF\ApiProblem\ApiProblemStrategy'  => 'ZF\ApiProblem\View\ApiProblemStrategy',
    ],
    'factories' => [
        'ZF\ApiProblem\Listener\ApiProblemListener'             => 'ZF\ApiProblem\Factory\ApiProblemListenerFactory',
        'ZF\ApiProblem\Listener\RenderErrorListener'            => 'ZF\ApiProblem\Factory\RenderErrorListenerFactory',
        'ZF\ApiProblem\Listener\SendApiProblemResponseListener' => 'ZF\ApiProblem\Factory\SendApiProblemResponseListenerFactory',
        'ZF\ApiProblem\View\ApiProblemRenderer'                 => 'ZF\ApiProblem\Factory\ApiProblemRendererFactory',
        'ZF\ApiProblem\View\ApiProblemStrategy'                 => 'ZF\ApiProblem\Factory\ApiProblemStrategyFactory',
    ],
],
'view_manager' => [
    // Enable this in your application configuration in order to get full
    // exception stack traces in your API-Problem responses.
    'display_exceptions' => false,
],
```

ZF2 Events
----------

### Listeners

#### ZF\ApiProblem\Listener\ApiProblemListener

The `ApiProblemListener` attaches to three events in the MVC lifecycle:

- `MvcEvent::EVENT_DISPATCH` as a _shared_ listener on `Zend\Stdlib\DispatchableInterface` with a
  priority of `100`.
- `MvcEvent::EVENT_DISPATCH_ERROR` with a priority of `100`.
- `MvcEvent::EVENT_RENDER` with a priority of `1000`.

If the current `Accept` media type does not match the configured API-Problem media types (by
default, these are `application/json` and `application/*+json`), then this listener returns without
taking any action.

When this listener does take action, the purposes are threefold:

- Before dispatching, the `render_error_controllers` configuration value is consulted to determine
  if the `ZF\ApiProblem\Listener\RenderErrorListener` should be attached; see
  [RenderErrorListener](#rendererrorlistener) for more information.
- After dispatching, detects the type of response from the controller; if it is already an
  `ApiProblem` model, it continues without doing anything. If an exception was thrown during
  dispatch, it converts the response to an API-Problem response with some information from the
  exception.
- If a dispatch error occurred, and the `Accept` type is in the set defined for API-Problems, it
  attempts to cast the dispatch exception into an API-Problem response.

#### ZF\ApiProblem\Listener\RenderErrorListener

This listener is attached to `MvcEvent::EVENT_RENDER_ERROR` at priority `100`.  This listener is
conditionally attached by `ZF\ApiProblem\Listener\ApiProblemListener` for controllers that require
API Problem responses.  With a priority of `100`, this ensures that this listener runs before the
default ZF2 listener on this event. In cases when it does run, it will cast an exception into an
API-problem response.

#### `ZF\ApiProblem\Listener\SendApiProblemResponseListener`

This listener is attached to `SendResponseEvent::EVENT_SEND_RESPONSE` at priority `-500`.  The
primary purpose of this listener is, on detection of an API-Problem response, to send appropriate
headers and the problem details as the content body. If the `view_manager`'s `display_exceptions`
setting is enabled, the listener will determine if the API-Problem represents an application
exception, and, if so, inject the exception trace as part of the serialized response.

ZF2 Services
------------

### Event Services

- `ZF\ApiProblem\Listener\ApiProblemListener`
- `ZF\ApiProblem\Listener\RenderErrorListener`
- `ZF\ApiProblem\Listener\SendApiProblemResponseListener`

### View Services

#### ZF\ApiProblem\View\ApiProblemRenderer

This service extends the `JsonRenderer` service from the ZF2 MVC layer.  Its primary responsibility
is to decorate JSON rendering with the ability to optionally output stack traces.

#### ZF\ApiProblem\View\ApiProblemStrategy

This service is a view strategy that detects a `ZF\ApiProblem\View\ApiProblemModel`; when detected,
it selects the [ApiProblemRender](#zfapiproblemviewapiproblemrenderer), and injects the response
with a `Content-Type` header that contains the `application/problem+json` media type. This is
similar in nature to Zend Framework 2's `JsonStrategy`.

### Models

#### ZF\ApiProblem\ApiProblem

An instance of `ZF\ApiProblem\ApiProblem` serves the purpose of modeling the kind of problem that is
encountered.  An instance of `ApiProblem` is typically wrapped in an
[ApiProblemResponse](#zfapiproblemapiproblemresponse). Most information can be passed into the
constructor:

```php
class ApiProblem
{
    public function __construct(
        $status,
        $detail,
        $type = null,
        $title = null,
        array $additional = []
    ) {
        /* ... */
    }
}
```

For example:

```php
new ApiProblem(404, 'Entity not found');

// or

new ApiProblem(424, $exceptionInstance);
```

#### `ZF\ApiProblem\ApiProblemResponse`

An instance of `ZF\ApiProblem\ApiProblemResponse` can be returned from any controller service or ZF2
MVC event in order to short-circuit the MVC lifecycle and immediately return a response. When it
is, the response will be converted to the proper JSON structure for an API-Problem, and the
`Content-Type` header will be set to the `application/problem+json` media type.

For example:

```php
use Zend\Mvc\Controller\AbstractActionController;
use ZF\ApiProblem\ApiProblem;
use ZF\ApiProblem\ApiProblemResponse;

class MyController extends AbstractActionController
{
    /* ... */
    public function fetch($id)
    {
        $entity = $this->model->fetch($id);
        if (! $entity) {
            return new ApiProblemResponse(new ApiProblem(404, 'Entity not found'));
        }
        return $entity;
    }
}
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file, in reverse chronological order by release.

## 1.4.0 - TBD

### Added

- Nothing.

### Changed

- Nothing.

### Deprecated

- Nothing.

### Removed

- Nothing.

### Fixed

- Nothing.

## 1.3.1 - TBD

### Added

- Nothing.

### Changed

- Nothing.

### Deprecated

- Nothing.

### Removed

- Nothing.

### Fixed

- Nothing.

## 1.3.0 - 2018-05-08

### Added

- [#58](https://github.com/zfcampus/zf-api-problem/pull/58) adds support for PHP 7.2.

### Changed

- Nothing.

### Deprecated

- Nothing.

### Removed

- [#58](https://github.com/zfcampus/zf-api-problem/pull/58) removes support for HHVM.

### Fixed

- Nothing.

## 1.2.3 - 2017-07-24

### Added

- Nothing.

### Deprecated

- Nothing.

### Removed

- Nothing.

### Fixed

- [#52](https://github.com/zfcampus/zf-api-problem/pull/52) and
  [#55](https://github.com/zfcampus/zf-api-problem/pull/55) update the
  `ApiProblemListener` and `RenderErrorListener` to check for either exceptions
  or PHP 7 `Throwable` instances when creating an `ApiProblem` to return.

## 1.2.2 - 2016-10-11

### Added

- Nothing.

### Deprecated

- Nothing.

### Removed

- Nothing.

### Fixed

- [#49](https://github.com/zfcampus/zf-api-problem/pull/49) fixes `ApiProblem`'s
  detail message marshaling when PHP 7 `Throwable` types are provided as the
  detail.

## 1.2.1 - 2016-07-07

### Added

- Nothing.

### Deprecated

- Nothing.

### Removed

- Nothing.

### Fixed

- [#47](https://github.com/zfcampus/zf-api-problem/pull/47) fixes an issue with
  the various factories introduced in 1.2.0; they had been updated to implement the
  zend-servicemanager v3 interfaces, which broke usage with v2. They are now
  written in a way that will work with both versions.

## 1.2.0 - 2016-07-06

### Added

- [#45](https://github.com/zfcampus/zf-api-problem/pull/45) adds support for PHP 7.
- [#44](https://github.com/zfcampus/zf-api-problem/pull/44) and
  [#45](https://github.com/zfcampus/zf-api-problem/pull/45) add support for
  version 3 components of Zend Framework.
- [#39](https://github.com/zfcampus/zf-api-problem/pull/39) adds the constant
  `ApiProblem::CONTENT_TYPE` for specifying the Content-Type of API Problem
  responses.

### Deprecated

- Nothing.

### Removed

- [#45](https://github.com/zfcampus/zf-api-problem/pull/45) removes support for
  PHP 5.6.
- [#45](https://github.com/zfcampus/zf-api-problem/pull/45) removes the
  `Module::getAutoloaderConfig()` implementation, as it was redundant in
  composer-based applications.

### Fixed

- [#44](https://github.com/zfcampus/zf-api-problem/pull/44) and
  [#45](https://github.com/zfcampus/zf-api-problem/pull/45) ensure that
  definition and attachment of the listener aggregates defined in the module
  will work with both v2 and v3 versions of zend-eventmanager.
- [#38](https://github.com/zfcampus/zf-api-problem/pull/38) fixes an issue
  whereby JSON encoding failed for included stack traces if they originated
  within a PHP stream.


src/ApiProblem.php
<?php
/**
 * @license   http://opensource.org/licenses/BSD-3-Clause BSD-3-Clause
 * @copyright Copyright (c) 2014-2016 Zend Technologies USA Inc. (http://www.zend.com)
 */

namespace ZF\ApiProblem;

use Exception;
use Throwable;
use ZF\ApiProblem\Exception\InvalidArgumentException;
use ZF\ApiProblem\Exception\ProblemExceptionInterface;

/**
 * Object describing an API-Problem payload.
 */
class ApiProblem
{
    /**
     * Content type for api problem response
     */
    const CONTENT_TYPE = 'application/problem+json';

    /**
     * Additional details to include in report.
     *
     * @var array
     */
    protected $additionalDetails = [];

    /**
     * URL describing the problem type; defaults to HTTP status codes.
     *
     * @var string
     */
    protected $type = 'http://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html';

    /**
     * Description of the specific problem.
     *
     * @var string|Exception|Throwable
     */
    protected $detail = '';

    /**
     * Whether or not to include a stack trace and previous
     * exceptions when an exception is provided for the detail.
     *
     * @var bool
     */
    protected $detailIncludesStackTrace = false;

    /**
     * HTTP status for the error.
     *
     * @var int
     */
    protected $status;

    /**
     * Normalized property names for overloading.
     *
     * @var array
     */
    protected $normalizedProperties = [
        'type' => 'type',
        'status' => 'status',
        'title' => 'title',
        'detail' => 'detail',
    ];

    /**
     * Status titles for common problems.
     *
     * @var array
     */
    protected $problemStatusTitles = [
        // CLIENT ERROR
        400 => 'Bad Request',
        401 => 'Unauthorized',
        402 => 'Payment Required',
        403 => 'Forbidden',
        404 => 'Not Found',
        405 => 'Method Not Allowed',
        406 => 'Not Acceptable',
        407 => 'Proxy Authentication Required',
        408 => 'Request Time-out',
        409 => 'Conflict',
        410 => 'Gone',
        411 => 'Length Required',
        412 => 'Precondition Failed',
        413 => 'Request Entity Too Large',
        414 => 'Request-URI Too Large',
        415 => 'Unsupported Media Type',
        416 => 'Requested range not satisfiable',
        417 => 'Expectation Failed',
        418 => 'I\'m a teapot',
        422 => 'Unprocessable Entity',
        423 => 'Locked',
        424 => 'Failed Dependency',
        425 => 'Unordered Collection',
        426 => 'Upgrade Required',
        428 => 'Precondition Required',
        429 => 'Too Many Requests',
        431 => 'Request Header Fields Too Large',
        // SERVER ERROR
        500 => 'Internal Server Error',
        501 => 'Not Implemented',
        502 => 'Bad Gateway',
        503 => 'Service Unavailable',
        504 => 'Gateway Time-out',
        505 => 'HTTP Version not supported',
        506 => 'Variant Also Negotiates',
        507 => 'Insufficient Storage',
        508 => 'Loop Detected',
        511 => 'Network Authentication Required',
    ];

    /**
     * Title of the error.
     *
     * @var string
     */
    protected $title;

    /**
     * Constructor.
     *
     * Create an instance using the provided information. If nothing is
     * provided for the type field, the class default will be used;
     * if the status matches any known, the title field will be selected
     * from $problemStatusTitles as a result.
     *
     * @param int    $status
     * @param string|Exception|Throwable $detail
     * @param string $type
     * @param string $title
     * @param array  $additional
     */
    public function __construct($status, $detail, $type = null, $title = null, array $additional = [])
    {
        if ($detail instanceof ProblemExceptionInterface) {
            if (null === $type) {
                $type = $detail->getType();
            }
            if (null === $title) {
                $title = $detail->getTitle();
            }
            if (empty($additional)) {
                $additional = $detail->getAdditionalDetails();
            }
        }

        // Ensure a valid HTTP status
        if (! is_numeric($status)
            || ($status < 100)
            || ($status > 599)
        ) {
            $status = 500;
        }

        $this->status = $status;
        $this->detail = $detail;
        $this->title = $title;

        if (null !== $type) {
            $this->type = $type;
        }

        $this->additionalDetails = $additional;
    }

    /**
     * Retrieve properties.
     *
     * @param string $name
     * @return mixed
     * @throws InvalidArgumentException
     */
    public function __get($name)
    {
        $normalized = strtolower($name);
        if (in_array($normalized, array_keys($this->normalizedProperties))) {
            $prop = $this->normalizedProperties[$normalized];

            return $this->{$prop};
        }

        if (isset($this->additionalDetails[$name])) {
            return $this->additionalDetails[$name];
        }

        if (isset($this->additionalDetails[$normalized])) {
            return $this->additionalDetails[$normalized];
        }

        throw new InvalidArgumentException(sprintf(
            'Invalid property name "%s"',
            $name
        ));
    }

    /**
     * Cast to an array.
     *
     * @return array
     */
    public function toArray()
    {
        $problem = [
            'type' => $this->type,
            'title' => $this->getTitle(),
            'status' => $this->getStatus(),
            'detail' => $this->getDetail(),
        ];
        // Required fields should always overwrite additional fields
        return array_merge($this->additionalDetails, $problem);
    }

    /**
     * Set the flag indicating whether an exception detail should include a
     * stack trace and previous exception information.
     *
     * @param bool $flag
     * @return ApiProblem
     */
    public function setDetailIncludesStackTrace($flag)
    {
        $this->detailIncludesStackTrace = (bool) $flag;

        return $this;
    }

    /**
     * Retrieve the API-Problem detail.
     *
     * If an exception was provided, creates the detail message from it;
     * otherwise, detail as provided is used.
     *
     * @return string
     */
    public function getDetail()
    {
        if ($this->detail instanceof Throwable || $this->detail instanceof Exception) {
            return $this->createDetailFromException();
        }

        return $this->detail;
    }

    /**
     * Retrieve the raw API-Problem detail.
     *
     * @return string|Exception|Throwable
     */
    public function getRawDetail()
    {
        return $this->detail;
    }

    /**
     * Retrieve the API-Problem HTTP status code.
     *
     * If an exception was provided, creates the status code from it;
     * otherwise, code as provided is used.
     *
     * @return string
     */
    public function getStatus()
    {
        if ($this->detail instanceof Throwable || $this->detail instanceof Exception) {
            $this->status = $this->createStatusFromException();
        }

        return $this->status;
    }

    /**
     * Retrieve the title.
     *
     * If the default $type is used, and the $status is found in
     * $problemStatusTitles, then use the matching title.
     *
     * If no title was provided, and the above conditions are not met, use the
     * string 'Unknown'.
     *
     * Otherwise, use the title provided.
     *
     * @return string
     */
    protected function getTitle()
    {
        if (null !== $this->title) {
            return $this->title;
        }

        if (null === $this->title
            && $this->type == 'http://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html'
            && array_key_exists($this->getStatus(), $this->problemStatusTitles)
        ) {
            return $this->problemStatusTitles[$this->status];
        }

        if ($this->detail instanceof Throwable || $this->detail instanceof Exception) {
            return get_class($this->detail);
        }

        if (null === $this->title) {
            return 'Unknown';
        }

        return $this->title;
    }

    /**
     * Create detail message from an exception.
     *
     * @return string
     */
    protected function createDetailFromException()
    {
        /** @var Exception|Throwable $e */
        $e = $this->detail;

        if (! $this->detailIncludesStackTrace) {
            return $e->getMessage();
        }

        $message = trim($e->getMessage());
        $this->additionalDetails['trace'] = $e->getTrace();

        $previous = [];
        $e = $e->getPrevious();
        while ($e) {
            $previous[] = [
                'code' => (int) $e->getCode(),
                'message' => trim($e->getMessage()),
                'trace' => $e->getTrace(),
            ];
            $e = $e->getPrevious();
        }
        if (count($previous)) {
            $this->additionalDetails['exception_stack'] = $previous;
        }

        return $message;
    }

    /**
     * Create HTTP status from an exception.
     *
     * @return int
     */
    protected function createStatusFromException()
    {
        /** @var Exception|Throwable $e */
        $e = $this->detail;
        $status = $e->getCode();

        if ($status) {
            return $status;
        }

        return 500;
    }
}
